dialogue-target/src/main/java/com/palantir/dialogue/DeserializerArgs.java

```diff
@@ -0,0 +1,108 @@
+
+public final class DeserializerArgs<T> {
+    private final TypeMarker<T> baseType;
+    private final TypeMarker<? extends T> expectedResultType;
+    private final ImmutableMap<String, TypeMarker<? extends T>> errorNameToTypeMarker;
+
+    private DeserializerArgs(
+            TypeMarker<T> baseType,
+            TypeMarker<? extends T> expectedResultType,
+            ImmutableMap<String, TypeMarker<? extends T>> map) {
+        this.baseType = baseType;
+        this.expectedResultType = expectedResultType;
+        this.errorNameToTypeMarker = map;
+    }
+
+    public static <T> Builder<T> builder() {
+        return new Builder<>();
+    }
+
+    public static final class Builder<T> {
+        private boolean buildInvoked = false;
+        private TypeMarker<T> baseType;
+        private boolean baseTypeSet = false;
+        private TypeMarker<? extends T> expectedResultType;
+        private boolean expectedResultSet = false;
+        private final Map<String, TypeMarker<? extends T>> errorNameToTypeMarker;
+
+        @SuppressWarnings("NullAway")
+        // We ensure that the baseType and expectedResultType are set before building.
+        private Builder() {
+            this.errorNameToTypeMarker = new HashMap<>();
+        }
+
+        public Builder<T> withBaseType(@Nonnull TypeMarker<T> base) {
+            checkNotBuilt();
+            this.baseType = Preconditions.checkNotNull(base, "base type must be non-null");
+            this.baseTypeSet = true;
+            return this;
+        }
+
+        public Builder<T> withExpectedResult(TypeMarker<? extends T> expectedResultT) {
+            checkNotBuilt();
+            this.expectedResultType =
+                    Preconditions.checkNotNull(expectedResultT, "expected result type must be non-null");
+            this.expectedResultSet = true;
+            return this;
+        }
+
+        public Builder<T> withErrorType(@Nonnull String errorName, @Nonnull TypeMarker<? extends T> errorType) {
+            checkNotBuilt();
+            this.errorNameToTypeMarker.put(
+                    Preconditions.checkNotNull(errorName, "error name must be non-null"),
+                    Preconditions.checkNotNull(errorType, "error type must be non-null"));
+            return this;
+        }
+
+        public DeserializerArgs<T> build() {
+            checkNotBuilt();
+            checkRequiredArgsSet();
+            buildInvoked = true;
+            return new DeserializerArgs<>(baseType, expectedResultType, ImmutableMap.copyOf(errorNameToTypeMarker));
+        }
+
+        private void checkNotBuilt() {
+            Preconditions.checkState(!buildInvoked, "Build has already been called");
+        }
+
+        private void checkRequiredArgsSet() {
+            Preconditions.checkState(baseTypeSet, "base type must be set");
+            Preconditions.checkState(expectedResultSet, "expected result type must be set");
+        }
+    }
+
+    public TypeMarker<? extends T> baseType() {
```

Comment:
Are there cases where this isn't `TypeMarker<T>` but a subtype of T?

Reply:
I don't think so. The `baseType` variable is also constrained to `TypeMarker<T>`. Updated.